File: rmd_chunks/chunks_as_scripts/dca-script.py
# ---- install ----- 

# install dcurves to perform DCA (first install package via pip)
# pip install dcurves
from dcurves import dca, plot_graphs

# install other packages used in this tutorial
# pip install pandas numpy statsmodels lifelines
import pandas as pd
import numpy as np
import statsmodels.api as sm
import lifelines

# ---- import_cancer ----- 

# No table for this chunk
df_cancer_dx = pd.read_csv(
    "https://raw.githubusercontent.com/ddsjoberg/dca-tutorial/main/data/df_cancer_dx.csv"
)

# ---- model ----- 

mod1 = sm.GLM.from_formula(
    "cancer ~ famhistory", data=df_cancer_dx, family=sm.families.Binomial()
)
mod1_results = mod1.fit()

print(mod1_results.summary())

# ---- dca_famhistory ----- 

dca_famhistory_df = dca(data=df_cancer_dx, outcome="cancer", modelnames=["famhistory"])

plot_graphs(plot_df=dca_famhistory_df, graph_type="net_benefit", y_limits=[-0.05, 0.2])

# ---- dca_famhistory2 ----- 

dca_famhistory2_df = dca(
    data=df_cancer_dx,
    outcome="cancer",
    modelnames=["famhistory"],
    thresholds=np.arange(0, 0.36, 0.01),
)

plot_graphs(plot_df=dca_famhistory2_df, graph_type="net_benefit", y_limits=[-0.05, 0.2])

# ---- model_multi ----- 

mod2 = sm.GLM.from_formula(
    "cancer ~ marker + age + famhistory",
    data=df_cancer_dx,
    family=sm.families.Binomial(),
)
mod2_results = mod2.fit()

print(mod2_results.summary())

df_cancer_dx["cancerpredmarker"] = mod2_results.predict(df_cancer_dx)

# ---- dca_multi ----- 

# Run dca on multivariable model
dca_multi_df = dca(
    data=df_cancer_dx,
    outcome="cancer",
    modelnames=["famhistory", "cancerpredmarker"],
    thresholds=np.arange(0, 0.36, 0.01),
)

plot_graphs(plot_df=dca_multi_df, y_limits=[-0.05, 0.2], graph_type="net_benefit")

# ---- dca_formatting ----- 

dca_formatting_df = dca(
    data=df_cancer_dx,
    outcome="cancer",
    modelnames=["cancerpredmarker"],
    thresholds=np.arange(0, 0.26, 0.01),
)

plot_graphs(
    plot_df=dca_formatting_df,
    graph_type="net_benefit",
    y_limits=[-0.05, 0.15],
    color_names=["blue", "red", "green"],
    linestyles=["--", "-", "--"],
    linewidths=[0.75, 3, 2],
)

# ---- dca_legend_off ----- 

# Turn off the legend for a cleaner publication-ready figure
dca_legend_off_df = dca(
    data=df_cancer_dx,
    outcome="cancer",
    modelnames=["cancerpredmarker"],
    thresholds=np.arange(0, 0.26, 0.01),
)

plot_graphs(
    plot_df=dca_legend_off_df,
    graph_type="net_benefit",
    y_limits=[-0.05, 0.15],
    color_names=["blue", "red", "green"],
    linestyles=["--", "-", "--"],
    linewidths=[0.75, 3, 2],
    show_legend=False,
)

# ---- dca_create_low_incidence ----- 

np.random.seed(123)

df_cancer_dx_low = df_cancer_dx.copy()
u = np.random.rand(len(df_cancer_dx_low))

# copy → cancer_temp, then blank out 90 % of cancers
df_cancer_dx_low["cancer_temp"] = df_cancer_dx_low["cancer"]
mask_na = (df_cancer_dx_low["cancer"] == 1) & (u < 0.9)
df_cancer_dx_low.loc[mask_na, "cancer_temp"] = np.nan

# drop rows missing in *either* outcome or predictor
df_for_dca = df_cancer_dx_low.dropna(subset=["cancer_temp", "cancerpredmarker"])

dca_low_incidence_df = dca(
    data=df_for_dca,
    outcome="cancer_temp",
    modelnames=["cancerpredmarker"],
    thresholds=np.arange(0, 0.26, 0.01),
)

plot_graphs(
    plot_df=dca_low_incidence_df,
    graph_type="net_benefit",
    color_names=["blue", "red", "green"],
)

# ---- dca_adjust_axes ----- 

dca_low_incidence_small = dca(
    data=df_for_dca,
    outcome="cancer_temp",
    modelnames=["cancerpredmarker"],
    thresholds=np.arange(0, 0.05, 0.01),
)

plot_graphs(
    plot_df=dca_low_incidence_small,
    graph_type="net_benefit",
    y_limits=[-0.005, 0.0225],
    color_names=["blue", "red", "green"],
)

# ---- dca_smooth ----- 


plot_graphs(
    plot_df=dca_multi_df,
    y_limits=[-0.025, 0.15],
    graph_type="net_benefit",
    smooth_frac=0.5,  # Set the smoothing fraction to 0.5
)

# ---- dca_smooth2 ----- 

dca_smooth2_df = dca(
    data=df_cancer_dx,
    outcome="cancer",
    modelnames=["cancerpredmarker", "famhistory", "risk_group"],
    thresholds=np.arange(0, 0.36, 0.05),
    models_to_prob=["risk_group"],  # Specify risk_group is not a probability
)

plot_graphs(
    plot_df=dca_smooth2_df,
    graph_type="net_benefit",
    smooth_frac=0,
    y_limits=[-0.025, 0.15],
)

# ---- pub_model ----- 

df_cancer_dx["logodds_brown"] = (
    0.75 * df_cancer_dx["famhistory"] + 0.26 * df_cancer_dx["age"] - 17.5
)
df_cancer_dx["phat_brown"] = np.exp(df_cancer_dx["logodds_brown"]) / (
    1 + np.exp(df_cancer_dx["logodds_brown"])
)

dca_pub_model_df = dca(
    data=df_cancer_dx,
    outcome="cancer",
    modelnames=["phat_brown"],
    thresholds=np.arange(0, 0.36, 0.01),
)

plot_graphs(
    plot_df=dca_pub_model_df,
    y_limits=[-0.05, 0.15],
    graph_type="net_benefit",
    smooth_frac=0.5,
    color_names=["blue", "red", "green"],
)

# ---- joint ----- 

df_cancer_dx["high_risk"] = np.where(df_cancer_dx["risk_group"] == "high", 1, 0)

df_cancer_dx["joint"] = np.where(
    (df_cancer_dx["risk_group"] == "high") | (df_cancer_dx["cancerpredmarker"] > 0.15),
    1,
    0,
)

df_cancer_dx["conditional"] = np.where(
    (df_cancer_dx["risk_group"] == "high")
    | (
        (df_cancer_dx["risk_group"] == "intermediate")
        & (df_cancer_dx["cancerpredmarker"] > 0.15)
    ),
    1,
    0,
)

# ---- dca_joint ----- 

dca_joint_df = dca(
    data=df_cancer_dx,
    outcome="cancer",
    modelnames=["high_risk", "joint", "conditional"],
    thresholds=np.arange(0, 0.36, 0.01),
)

plot_graphs(
    plot_df=dca_joint_df,
    graph_type="net_benefit",
    y_limits=[-0.05, 0.15],
    color_names=["green", "cyan", "violet", "red", "gold"],
)

# ---- dca_table ----- 

dca_table_df = dca(
    data=df_cancer_dx,
    outcome="cancer",
    modelnames=["marker"],
    models_to_prob=["marker"],
    thresholds=np.arange(0.05, 0.36, 0.15),
)

print("\n", dca_table_df[["model", "threshold", "net_benefit"]])

# ---- dca_intervention ----- 

dca_intervention_df = dca(
    data=df_cancer_dx,
    outcome="cancer",
    modelnames=["marker"],
    thresholds=np.arange(0.05, 0.36, 0.01),
    models_to_prob=["marker"],
)

plot_graphs(
    plot_df=dca_intervention_df,
    graph_type="net_intervention_avoided",
    smooth_frac=0.2,
    color_names=["blue", "red", "green"],
)

# ---- import_ttcancer ----- 

df_time_to_cancer_dx = pd.read_csv(
    "https://raw.githubusercontent.com/ddsjoberg/dca-tutorial/main/data/df_time_to_cancer_dx.csv"
)

# ---- coxph ----- 

cph = lifelines.CoxPHFitter()
cph.fit(
    df=df_time_to_cancer_dx,
    duration_col="ttcancer",
    event_col="cancer",
    formula="age + famhistory + marker",
)

cph_pred_vals = cph.predict_survival_function(
    df_time_to_cancer_dx[["age", "famhistory", "marker"]], times=[1.5]
)

df_time_to_cancer_dx["pr_failure18"] = [1 - val for val in cph_pred_vals.iloc[0, :]]

# ---- stdca_coxph ----- 

stdca_coxph_results = dca(
    data=df_time_to_cancer_dx,
    outcome="cancer",
    modelnames=["pr_failure18"],
    thresholds=np.arange(0, 0.51, 0.01),
    time=1.5,
    time_to_outcome_col="ttcancer",
)

plot_graphs(
    plot_df=stdca_coxph_results,
    graph_type="net_benefit",
    y_limits=[-0.05, 0.25],
    smooth_frac=0.5,
    color_names=["blue", "red", "green"],
)

# ---- stdca_cmprsk ----- 

# Python library doesn't support competing risks

# ---- import_case_control ----- 

df_case_control = pd.read_csv(
    "https://raw.githubusercontent.com/ddsjoberg/dca-tutorial/main/data/df_cancer_dx_case_control.csv"
)

# Summarize Data With Column Medians
# drop 'patientid', then group and take medians of numeric columns only
medians = (
    df_case_control.drop(columns="patientid")
    .groupby("casecontrol")
    .median(numeric_only=True)
)
print(medians)

# ---- dca_case_control ----- 

dca_case_control_df = dca(
    data=df_case_control,
    outcome="casecontrol",
    modelnames=["cancerpredmarker"],
    prevalence=0.20,
    thresholds=np.arange(0, 0.51, 0.01),
)

plot_graphs(
    plot_df=dca_case_control_df,
    graph_type="net_benefit",
    y_limits=[-0.05, 0.25],
    smooth_frac=0.1,
)

# ---- dca_harm_simple ----- 

dca_harm_simple_df = dca(
    data=df_cancer_dx,
    outcome="cancer",
    modelnames=["marker"],
    thresholds=np.arange(0, 0.36, 0.01),
    harm={"marker": 0.0333},
    models_to_prob=["marker"],
)

plot_graphs(
    plot_df=dca_harm_simple_df,
    graph_type="net_benefit",
    y_limits=[-0.05, 0.15],
    color_names=["blue", "red", "green"],
    smooth_frac=0.2,
)

# ---- dca_harm ----- 

harm_marker = 0.0333
harm_conditional = (df_cancer_dx["risk_group"] == "intermediate").mean() * harm_marker

dca_harm_df = dca(
    data=df_cancer_dx,
    outcome="cancer",
    modelnames=["risk_group"],
    models_to_prob=["risk_group"],
    thresholds=np.arange(0, 0.36, 0.01),
    harm={"risk_group": harm_conditional},
)

plot_graphs(plot_df=dca_harm_df)

# ---- cross_validation ----- 

# Load dependencies for cross validation
import random  # Library to generate random seed
from sklearn.model_selection import (
    RepeatedKFold,
)  # Cross validation data selection and segregation function

# Set seed for random processes
random.seed(112358)

# Load simulation data
df_cancer_dx = pd.read_csv(
    "https://raw.githubusercontent.com/ddsjoberg/dca-tutorial/main/data/df_cancer_dx.csv"
)

# Define the formula (make sure the column names in your DataFrame match these)
formula = "cancer ~ marker + age + famhistory"

# Create cross-validation object
rkf = RepeatedKFold(n_splits=10, n_repeats=1, random_state=112358)

# Placeholder for predictions
cv_predictions = []

# Perform cross-validation
for train_index, test_index in rkf.split(df_cancer_dx):
    # Split data into training and test sets
    train = df_cancer_dx.iloc[train_index]
    test = df_cancer_dx.iloc[test_index].copy()  # ← make an explicit copy

    # Fit the model
    model = sm.Logit.from_formula(formula, data=train).fit(disp=0)

    # Make predictions on the test set
    test["cv_prediction"] = model.predict(test)

    # Store predictions
    cv_predictions.append(test[["patientid", "cv_prediction"]])

# Concatenate predictions from all folds
df_predictions = pd.concat(cv_predictions)

# Calculate mean prediction per patient
df_mean_predictions = (
    df_predictions.groupby("patientid")["cv_prediction"].mean().reset_index()
)

# Join with original data
df_cv_pred = pd.merge(df_cancer_dx, df_mean_predictions, on="patientid", how="left")

# Decision curve analysis
df_dca_cv = dca(
    data=df_cv_pred,
    modelnames=["cv_prediction"],
    outcome="cancer",
    thresholds=np.arange(0, 0.36, 0.01),
)

# Plot DCA curves
plot_graphs(
    plot_df=df_dca_cv,
    graph_type="net_benefit",
    y_limits=[-0.01, 0.15],
    color_names=["blue", "red", "green"],
    smooth_frac=0.5,
)

File: rmd_chunks/chunks_as_scripts/dca-script.R
# ---- install ----- 
# install dcurves to perform DCA from CRAN
install.packages("dcurves")

# install other packages used in this tutorial
install.packages(
  c("tidyverse", "survival", "gt", "broom",
    "gtsummary", "rsample", "labelled")
)

# load packages
library(dcurves)
library(tidyverse)
library(gtsummary)

# ---- import_cancer ----- 
# import data
df_cancer_dx <-
  readr::read_csv(
    file = "https://raw.githubusercontent.com/ddsjoberg/dca-tutorial/main/data/df_cancer_dx.csv"
  ) %>%
  # assign variable labels. these labels will be carried through in the `dca()` output
  labelled::set_variable_labels(
    patientid = "Patient ID",
    cancer = "Cancer Diagnosis",
    risk_group = "Risk Group",
    age = "Patient Age",
    famhistory = "Family History",
    marker = "Marker",
    cancerpredmarker = "Prediction Model"
  )

# summarize data
df_cancer_dx %>%
  select(-patientid) %>%
  tbl_summary(type = all_dichotomous() ~ "categorical")

# ---- model ----- 
# build logistic regression model
mod1 <- glm(cancer ~ famhistory, data = df_cancer_dx, family = binomial)

# model summary
mod1_summary <- tbl_regression(mod1, exponentiate = TRUE)
mod1_summary

# ---- dca_famhistory ----- 
dca(cancer ~ famhistory, data = df_cancer_dx) %>%
  plot()

# ---- dca_famhistory2 ----- 
dca(cancer ~ famhistory,
    data = df_cancer_dx,
    thresholds = seq(0, 0.35, 0.01)
) %>%
  plot()

# ---- model_multi ----- 
# build multivariable logistic regression model
mod2 <- glm(cancer ~ marker + age + famhistory, data = df_cancer_dx, family = binomial)

# summarize model
tbl_regression(mod2, exponentiate = TRUE)

# add predicted values from model to data set
df_cancer_dx <-
  df_cancer_dx %>%
  mutate(
    cancerpredmarker =
      broom::augment(mod2, type.predict = "response") %>%
      pull(".fitted")
  )

# ---- dca_multi ----- 
dca(cancer ~ famhistory + cancerpredmarker,
    data = df_cancer_dx,
    thresholds = seq(0, 0.35, 0.01),
    label = list(cancerpredmarker = "Prediction Model")
) %>%
  plot(smooth = FALSE)

# ---- dca_smooth ----- 
dca(cancer ~ famhistory + cancerpredmarker,
    data = df_cancer_dx,
    thresholds = seq(0, 0.35, 0.01),
    label = list(cancerpredmarker = "Prediction Model")
) %>%
  plot(smooth = TRUE)

# ---- dca_smooth2 ----- 
dca(cancer ~ famhistory + cancerpredmarker,
    data = df_cancer_dx,
    thresholds = seq(0, 0.35, 0.05),
    label = list(cancerpredmarker = "Prediction Model")
) %>%
  plot(smooth = FALSE)

# ---- pub_model ----- 
# Use the coefficients from the Brown model
df_cancer_dx <-
  df_cancer_dx %>%
  mutate(
    logodds_brown = 0.75 * famhistory + 0.26 * age - 17.5,
    phat_brown = exp(logodds_brown) / (1 + exp(logodds_brown))
  )

# Run the decision curve
dca(cancer ~ phat_brown,
    data = df_cancer_dx,
    thresholds = seq(0, 0.35, 0.01),
    label = list(phat_brown = "Brown Model")
) %>%
  plot(smooth = TRUE)

# ---- joint ----- 
# Create a variable for the strategy of treating only high risk patients
df_cancer_dx <-
  df_cancer_dx %>%
  mutate(
    # This will be 1 for treat and 0 for don’t treat
    high_risk = ifelse(risk_group == "high", 1, 0),
    # Treat based on Joint Approach
    joint = ifelse(risk_group == "high" | cancerpredmarker > 0.15, 1, 0),
    # Treat based on Conditional Approach
    conditional =
      ifelse(risk_group == "high" | (risk_group == "intermediate" & cancerpredmarker > 0.15), 1, 0)
  )

# ---- dca_joint ----- 
dca(cancer ~ high_risk + joint + conditional,
    data = df_cancer_dx,
    thresholds = seq(0, 0.35, 0.01),
    label = list(
      high_risk = "High Risk",
      joint = "Joint Test",
      conditional = "Conditional Approach"
    )
) %>%
  plot(smooth = TRUE)

# ---- dca_harm_simple ----- 
# Run the decision curve
dca(cancer ~ marker,
    data = df_cancer_dx,
    thresholds = seq(0, 0.35, 0.01),
    as_probability = "marker",
    harm = list(marker = 0.0333)
) %>%
  plot(smooth = TRUE)

# ---- dca_harm ----- 
# the harm of measuring the marker is stored in a scalar
harm_marker <- 0.0333
# in the conditional test, only patients at intermediate risk
# have their marker measured
# harm of the conditional approach is proportion of patients who have the marker
# measured multiplied by the harm
harm_conditional <- mean(df_cancer_dx$risk_group == "intermediate") * harm_marker

# Run the decision curve
dca(cancer ~ risk_group,
    data = df_cancer_dx,
    thresholds = seq(0, 0.35, 0.01),
    as_probability = "risk_group",
    harm = list(risk_group = harm_conditional)
) %>%
  plot(smooth = TRUE)


# ---- dca_table ----- 
dca(cancer ~ marker,
    data = df_cancer_dx,
    as_probability = "marker",
    thresholds = seq(0.05, 0.35, 0.15)
) %>%
  as_tibble() %>%
  select(label, threshold, net_benefit) %>%
  gt::gt() %>%
  gt::fmt_percent(columns = threshold, decimals = 0) %>%
  gt::fmt(columns = net_benefit, fns = function(x) style_sigfig(x, digits = 3)) %>%
  gt::cols_label(
    label = "Strategy",
    threshold = "Decision Threshold",
    net_benefit = "Net Benefit"
  ) %>%
  gt::cols_align("left", columns = label)

# ---- dca_intervention ----- 
dca(cancer ~ marker,
    data = df_cancer_dx,
    as_probability = "marker",
    thresholds = seq(0.05, 0.35, 0.01),
    label = list(marker = "Marker")
) %>%
  net_intervention_avoided() %>%
  plot(smooth = TRUE)

# ---- import_ttcancer ----- 
# import data
df_time_to_cancer_dx <-
  readr::read_csv(
    file = "https://raw.githubusercontent.com/ddsjoberg/dca-tutorial/main/data/df_time_to_cancer_dx.csv"
  ) %>%
  # assign variable labels. these labels will be carried through in the `dca()` output
  labelled::set_variable_labels(
    patientid = "Patient ID",
    cancer = "Cancer Diagnosis",
    ttcancer = "Years to Diagnosis/Censor",
    risk_group = "Risk Group",
    age = "Patient Age",
    famhistory = "Family History",
    marker = "Marker",
    cancerpredmarker = "Prediction Model",
    cancer_cr = "Cancer Diagnosis Status"
  )

# ---- coxph ----- 
# Load survival library
library(survival)

# Run the cox model
coxmod <- coxph(Surv(ttcancer, cancer) ~ age + famhistory + marker, data = df_time_to_cancer_dx)

df_time_to_cancer_dx <-
  df_time_to_cancer_dx %>%
  mutate(
    pr_failure18 =
      1 - summary(survfit(coxmod, newdata = df_time_to_cancer_dx), times = 1.5)$surv[1, ]
  )

# ---- stdca_coxph ----- 
dca(Surv(ttcancer, cancer) ~ pr_failure18,
    data = df_time_to_cancer_dx,
    time = 1.5,
    thresholds = seq(0, 0.5, 0.01),
    label = list(pr_failure18 = "Prediction Model")
) %>%
  plot(smooth = TRUE)


# ---- stdca_cmprsk ----- 
# status variable must be a factor with first level coded as 'censor',
# and the second level the outcome of interest.
df_time_to_cancer_dx <-
  df_time_to_cancer_dx %>%
  mutate(
    cancer_cr =
      factor(cancer_cr,
             levels = c("censor", "diagnosed with cancer", "dead other causes")
      )
  )

dca(Surv(ttcancer, cancer_cr) ~ pr_failure18,
    data = df_time_to_cancer_dx,
    time = 1.5,
    thresholds = seq(0, 0.5, 0.01),
    label = list(pr_failure18 = "Prediction Model")
) %>%
  plot(smooth = TRUE)

# ---- import_case_control ----- 
# import data
df_cancer_dx_case_control <-
  readr::read_csv(
    file = "https://raw.githubusercontent.com/ddsjoberg/dca-tutorial/main/data/df_cancer_dx_case_control.csv"
  ) %>%
  # assign variable labels. these labels will be carried through in the `dca()` output
  labelled::set_variable_labels(
    patientid = "Patient ID",
    casecontrol = "Case-Control Status",
    risk_group = "Risk Group",
    age = "Patient Age",
    famhistory = "Family History",
    marker = "Marker",
    cancerpredmarker = "Prediction Model"
  )

# summarize data
df_cancer_dx_case_control %>%
  select(-patientid) %>%
  tbl_summary(
    by = casecontrol,
    type = all_dichotomous() ~ "categorical"
  ) %>%
  modify_spanning_header(all_stat_cols() ~ "**Case-Control Status**")

# ---- dca_case_control ----- 
dca(casecontrol ~ cancerpredmarker,
    data = df_cancer_dx_case_control,
    prevalence = 0.20,
    thresholds = seq(0, 0.5, 0.01)
) %>%
  plot(smooth = TRUE)

# ---- cross_validation ----- 
# set seed for random process
set.seed(112358)

formula = cancer ~ marker + age + famhistory
dca_thresholds = seq(0,0.36, 0.01)

# create a 10-fold cross validation set, 1 repeat which is the base case, change to suit your use case
cross_validation_samples <- rsample::vfold_cv(df_cancer_dx, v = 10, repeats = 1)

df_crossval_predictions <- 
  cross_validation_samples %>% 
  # for each cut of the data, build logistic regression on the 90% (analysis set)
  rowwise() %>% 
  mutate(
    # build regression model on analysis set
    glm_analysis =
      glm(formula = formula,
          data = rsample::analysis(splits),
          family = binomial
      ) %>%
      list(),
    # get predictions for assessment set
    df_assessment =
      broom::augment(
        glm_analysis,
        newdata = rsample::assessment(splits),
        type.predict = "response"
      ) %>%
      list()
  ) %>%
  ungroup() %>%
  # pool results from the 10-fold cross validation
  pull(df_assessment) %>%
  bind_rows() %>% # Concatenate all cross validation predictions
  group_by(patientid) %>%
  summarise(cv_pred = mean(.fitted), .groups = "drop") %>% # Generate mean prediction per patient
  ungroup()

df_cv_pred <-
  df_cancer_dx %>%
  left_join(
    df_crossval_predictions,
    by = 'patientid'
  )

dcurves::dca( # calculate net benefit scores on mean cross validation predictions
  data = df_cv_pred,
  formula = cancer ~ cv_pred,
  thresholds = dca_thresholds,
  label = list(
    cv_pred = "Cross-validated Prediction Model"
  ))$dca |> 
  # plot cross validated net benefit values
  ggplot(aes(x = threshold, y = net_benefit, color = label)) +
  stat_smooth(method = "loess", se = FALSE, formula = "y ~ x", span = 0.2) +
  coord_cartesian(ylim = c(-0.014, 0.14)) +
  scale_x_continuous(labels = scales::percent_format(accuracy = 1)) +
  labs(x = "Threshold Probability", y = "Net Benefit", color = "") +
  theme_bw()
